bolts/models/dann.py

```diff
@@ -0,0 +1,195 @@
+        )
+
+        tm_acc_s = self.val_acc_s if stage == "val" else self.test_acc_s
+        tm_acc_y = self.val_acc_y if stage == "val" else self.test_acc_y
+        acc_s = tm_acc_s.compute().item()
+        acc_y = tm_acc_y.compute().item()
+        results_dict = {f"{stage}/acc_s": acc_s, f"{stage}/acc_y": acc_y}
+        results_dict.update({f"{stage}/{k}": v for k, v in results.items()})
+        return results_dict
+
+    def _inference_step(self, batch: DataBatch, stage: Stage) -> Dict[str, Tensor]:
+        _s, _y = self.forward(batch.x)
+        target_s = batch.s.view(-1, 1).float()
+        loss_adv = self._loss_adv_fn(_s, target_s)
+        target_y = batch.y.view(-1, 1).float()
+        loss_clf = self._loss_clf_fn(_y, target_y)
+        loss = loss_adv + loss_clf
+        tm_acc_s = self.val_acc_s if stage == "val" else self.test_acc_s
+        tm_acc_y = self.val_acc_y if stage == "val" else self.test_acc_y
+
+        target_y = batch.y.view(-1, 1).long()
+        y_acc = tm_acc_y(_y >= 0, target_y)
+
+        target_s = batch.s.view(-1, 1).long()
+        s_acc = tm_acc_s(_s >= 0, target_s)
+        self.log_dict(
+            {
+                f"{stage}/loss": loss.item(),
+                f"{stage}/loss_adv": loss_adv.item(),
+                f"{stage}/loss_clf": loss_clf.item(),
+                f"{stage}/acc_s": s_acc,
+                f"{stage}/acc_y": y_acc,
+            }
+        )
+        return {"y": batch.y, "s": batch.s, "preds": _y.sigmoid().round().squeeze(-1)}
+
+    def reset_parameters(self) -> None:
+        """Reset the models."""
+        self.adv.apply(self._maybe_reset_parameters)
+        self.enc.apply(self._maybe_reset_parameters)
+        self.clf.apply(self._maybe_reset_parameters)
+
+    @implements(pl.LightningModule)
+    def configure_optimizers(
+        self,
+    ) -> optim.Optimizer:
+        return optim.AdamW(
+            self.parameters(),
+            lr=self.learning_rate,
+            weight_decay=self.weight_decay,
+        )
+
+    @implements(pl.LightningModule)
+    def test_epoch_end(self, output_results: List[Dict[str, Tensor]]) -> None:
+        results_dict = self._inference_epoch_end(output_results=output_results, stage="test")
+        self.log_dict(results_dict)
+
+    @implements(pl.LightningModule)
+    def test_step(self, batch: DataBatch, batch_idx: int) -> Dict[str, Tensor]:
+        return self._inference_step(batch=batch, stage="test")
+
+    @implements(pl.LightningModule)
+    def training_step(self, batch: DataBatch, batch_idx: int) -> Tensor:
+        _s, _y = self.forward(batch.x)
+        target_s = batch.s.view(-1, 1).float()
+        loss_adv = self._loss_adv_fn(_s, target_s)
+        target_y = batch.y.view(-1, 1).float()
+        loss_clf = self._loss_clf_fn(_y, target_y)
+        loss = loss_adv + loss_clf
+
+        target_s = batch.s.view(-1, 1).long()
+        acc_s = self.train_acc_s(_s >= 0, target_s)
```

Comment:
I think there should probably be a wrapper around Accuracy to save us calling >= 0 all the time